test/system/src/test/java/io/pravega/test/system/ReaderCheckpointTest.java

```diff
@@ -0,0 +1,242 @@
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+import java.io.Serializable;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.IntSummaryStatistics;
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.Timeout;
+import org.junit.runner.RunWith;
+
+import io.pravega.client.ClientFactory;
+import io.pravega.client.admin.ReaderGroupManager;
+import io.pravega.client.admin.StreamManager;
+import io.pravega.client.stream.Checkpoint;
+import io.pravega.client.stream.EventRead;
+import io.pravega.client.stream.EventStreamReader;
+import io.pravega.client.stream.EventStreamWriter;
+import io.pravega.client.stream.EventWriterConfig;
+import io.pravega.client.stream.ReaderConfig;
+import io.pravega.client.stream.ReaderGroup;
+import io.pravega.client.stream.ReaderGroupConfig;
+import io.pravega.client.stream.ReinitializationRequiredException;
+import io.pravega.client.stream.ScalingPolicy;
+import io.pravega.client.stream.StreamConfiguration;
+import io.pravega.client.stream.impl.JavaSerializer;
+import io.pravega.common.concurrent.ExecutorServiceHelpers;
+import io.pravega.test.system.framework.Environment;
+import io.pravega.test.system.framework.SystemTestRunner;
+import io.pravega.test.system.framework.services.BookkeeperService;
+import io.pravega.test.system.framework.services.PravegaControllerService;
+import io.pravega.test.system.framework.services.PravegaSegmentStoreService;
+import io.pravega.test.system.framework.services.Service;
+import io.pravega.test.system.framework.services.ZookeeperService;
+import lombok.Cleanup;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+@RunWith(SystemTestRunner.class)
+public class ReaderCheckpointTest {
+
+    private static final long READ_TIMEOUT = SECONDS.toMillis(30);
+    private static final int RANDOM_SUFFIX = new Random().nextInt(Integer.MAX_VALUE);
+    private static final String SCOPE = "scope" + RANDOM_SUFFIX;
+    private static final String STREAM = "checkPointTestStream";
+    private static final String READER_GROUP_NAME = "checkpointTest" + RANDOM_SUFFIX;
+
+    @Rule
+    public Timeout globalTimeout = Timeout.seconds(7 * 60);
+
+    private final ReaderConfig readerConfig = ReaderConfig.builder().build();
+    private final ReaderGroupConfig readerGroupConfig = ReaderGroupConfig.builder().build();
+    private final ScheduledExecutorService executor = ExecutorServiceHelpers.newScheduledThreadPool(1, "checkPointExecutor");
+    private final StreamConfiguration streamConfig = StreamConfiguration.builder()
```

Comment:
I suggest we use multiple segments to stress the mechanism.

Reply:
Updated to use multiple segments.